bundles/org.openhab.binding.magentatv/src/main/java/org/openhab/binding/magentatv/internal/handler/MagentaTVHandler.java

```diff
@@ -0,0 +1,692 @@
+
+    /**
+     * If userID is empty and credentials are given the Telekom OAuth service is
+     * used to query the userID
+     *
+     * @throws MagentaTVException
+     */
+    private void authenticateUser() throws MagentaTVException {
+        String userId = config.getUserID();
+        if (userId.isEmpty()) {
+            // run OAuth authentication, this finally provides the userID
+            userId = control.authenticateUser(config.getAccountName(), config.getAccountPassword());
+
+            // Update thing configuration (persistent) - remove credentials, add userID
+            Configuration configuration = this.getConfig();
+            configuration.remove(PROPERTY_ACCT_NAME);
+            configuration.remove(PROPERTY_ACCT_PWD);
+            configuration.remove(PROPERTY_USERID);
+            configuration.put(PROPERTY_ACCT_NAME, "");
+            configuration.put(PROPERTY_ACCT_PWD, "");
+            configuration.put(PROPERTY_USERID, userId);
+            this.updateConfiguration(configuration);
+            config.setAccountName("");
+            config.setAccountPassword("");
+        } else {
+            logger.debug("{}: Skip OAuth, use existing userID {}", thingId, config.getUserID());
+        }
+        if (!userId.isEmpty()) {
+            config.setUserID(userId);
+        } else {
+            logger.warn("{}: Unable to obtain userId from OAuth", thingId);
+        }
+    }
+
+    /**
+     * Update thing status
+     *
+     * @param mode new thing status
+     * @return ON = power on, OFF=power off
+     */
+    public void setOnlineState(ThingStatus newStatus, String errorMessage) {
+        ThingStatus status = this.getThing().getStatus();
+        if (status != newStatus) {
+            if (newStatus == ThingStatus.ONLINE) {
+                updateStatus(newStatus);
+                updateState(CHANNEL_POWER, OnOffType.ON);
+            } else {
+                if (!errorMessage.isEmpty()) {
+                    logger.debug("{}: Communication Error - {}, switch Thing offline", thingId, errorMessage);
+                    updateStatus(newStatus, ThingStatusDetail.COMMUNICATION_ERROR, errorMessage);
+                } else {
+                    updateStatus(newStatus);
+                }
+                updateState(CHANNEL_POWER, OnOffType.OFF);
+            }
+        }
+    }
+
+    /**
+     * A wakeup of the MR was detected (e.g. UPnP received)
+     *
+     * @throws MagentaTVException
+     */
+    @Override
+    public void onWakeup(Map<String, String> discoveredProperties) throws MagentaTVException {
+        if ((this.getThing().getStatus() == ThingStatus.OFFLINE) || config.getVerificationCode().isEmpty()) {
+            // Device sent a UPnP discovery information, trigger to reconnect
+            connectReceiver();
+        } else {
+            logger.debug("{}: Refesh device status for {} (UDN={}", thingId, deviceName(), config.getUDN());
+            setOnlineState(ThingStatus.ONLINE, "");
```

Comment:
here's a good moment to set the power state to ON.

Reply:
that happens in setOnlineState()

Reply:
But here it would be much cleaner, since it is about sending state updates to channels and `setOnlineStatus` should be about setting the status of the Thing. Trying to reduce unexpected side-effects in methods here.